_posts/2020-01-31-Everything-you-need-to-know-about-ISP.md

```diff
@@ -0,0 +1,166 @@
+
+## What is an Interface?
+An Interface is a set of abstractions which implementing class must follow. Basically, you define the behaviour but not implement them.
+For e.g. You can define them in Java as below:
+```java
+interface Dog{
+  void bark();
+}
+
+class Poodle implements dog{
+  // Poodle class needs to implement dog otherwise we will get a compilation error
+  public void bark(){
+    
+  }
+}
+```
+
+## What is the Interface Segregation Principle (ISP)
+ISP states that **A Client should not be exposed to methods that they don't need**. Declaring methods in an interface which it doesn't need pollutes the interface and leads to the bulky interface also called a fat interface.
+
+## Why you need ISP in your code?
+Lets taken an example to understand this - Below is code for **Burgerpoint** where a customer can order burger, fries or a combo.
+``` java 
+/**
+ * IOrder.java
+ **/
+public interface IOrder {
+    void orderBurger(int quantity);
+    void orderFries(int fries);
+}
+```
+Since a customer can order fries or burger both, we decided to put both of them in a single interface. Now to support a burger only order, we have to throw an exception in **orderFries** method.
+```java
+/**
+ * BurgerOrder.java
+ **/
+public class BurgerOrder implements IOrder {
+    @Override
+    public void orderBurger(int quantity) {
+
+    }
+
+    @Override
+    public void orderFries(int fries) {
+        throw new UnsupportedOperationException("No fries in Burger only order");
+    }
+}
+
+// Similarly for FriesOrder.java
+```
+Similarly, for fries only order you also need to throw an exception in **orderBurger** method. This is not the only downside of this design, these classes will also have unwanted side effects, whenever you make changes to your abstraction. Let's say you decided to accept an order of fries in multiple units such as pounds or grams. In that case we most likely we have to add a parameter for the unit in orderFries method. This change will also affect the **BurgerOrder.java** even when it doesn't really support the ordering of fries. 
+
+By not following the ISP, we face the following problems in our code:
+1. We have unwanted methods in client classes
+2. Since we need to explicitly handle access to methods which client does not need, maintenance has become harder because of the side effect of ISP violation which will only be going to increase for a large codebase
+3. Availability of methods which client does not use or need confuses the client as well
+4. We will also see further how ISP violation will lead to violation of other SOLID principles as well.
+
+## Code smells for ISP violations and how to fix them
+Whether working solo or in larger teams, it helps to identify problems earlier in code, here are some code smells which could possibly indicate ISP violation:
+
+1. A bulky interface    
+A bulky interface might be a good indicator for ISP violation, although it's not always the case
```

Comment:
I have added the definition for same. This should have meant that we should ignore ISP violation when cohesion makes sense since we already have another heading for same, I have removed this line